### PR TITLE
Move controllers and PID objects to a real time update system

### PR DESCRIPTION
**Pull request by lthall**
This PR changes the PID objects, internal filters, Attitude Control, Position Control, and Waypoint Navigation to update based on the measured loop time rather than an aproximated regular time. This addresses serious control issues that present themselves when the autopilot becomes overloaded and the updates drop below the expected rate.

This PR has the added benifit of reducing noise caused by impefectly regular updates at all levels of the control system.  This noise is caused by incorect calculation of intergrals and derivitives due to innacurate values of time steps.
Filter performance is improved over all operating conditions resulting in more constant magnitude and phase delay and lower noise.

This PR should also remove the flow of control error caused by overloaded processor while also removing the negative impact this reduction in update rate has on control and navigation.

### CODE REVIEW

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp [hunk 1]
```diff
@@ -238,9 +238,9 @@ void AC_AttitudeControl::input_quaternion(Quaternion& attitude_desired_quat, Vec
         // When acceleration limiting and feedforward are enabled, the sqrt controller is used to compute an euler
         // angular velocity that will cause the euler angle to smoothly stop at the input angle with limited deceleration
         // and an exponential decay specified by _input_tc at the end.
-        _ang_vel_target.x = input_shaping_angle(wrap_PI(attitude_error_angle.x), _input_tc, get_accel_roll_max_radss(), _ang_vel_target.x, ang_vel_target.x, radians(_ang_vel_roll_max), _dt);
-        _ang_vel_target.y = input_shaping_angle(wrap_PI(attitude_error_angle.y), _input_tc, get_accel_pitch_max_radss(), _ang_vel_target.y, ang_vel_target.y, radians(_ang_vel_pitch_max), _dt);
-        _ang_vel_target.z = input_shaping_angle(wrap_PI(attitude_error_angle.z), _input_tc, get_accel_yaw_max_radss(), _ang_vel_target.z, ang_vel_target.z, radians(_ang_vel_yaw_max), _dt);
+        _ang_vel_target.x = input_shaping_angle(wrap_PI(attitude_error_angle.x), _input_tc, get_accel_roll_max_radss(), _ang_vel_target.x, ang_vel_target.x, radians(_ang_vel_roll_max), get_dt());
```
Comment (IamPete1) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:241]: you could do:
`const float _dt = get_dt();`

Because it uses the same variable name as the old internal variable the diff would be much less. Even if its a bit to odd to reuse the variable name I think pre-fetching like that would be more efficient.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp [hunk 2]
```diff
@@ -515,7 +516,7 @@ void AC_PosControl::init_xy_controller()
     init_ekf_xy_reset();
 
     // initialise z_controller time out
-    _last_update_xy_us = AP_HAL::micros64();
+    _last_update_xy_us = AP::ins().get_last_update_usec();
```
Comment (IamPete1) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp:519]: Why the switch to `last_update` here? I guess pos control cares about the time where as attitude control only cares about the change in time? There could be some case where the change in last update does not match the dt, at least micros should (ish).

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp [hunk 3]
```diff
@@ -1067,11 +1067,11 @@ float AC_AttitudeControl::get_althold_lean_angle_max_cd() const
 // Return roll rate step size in centidegrees/s that results in maximum output after 4 time steps
 float AC_AttitudeControl::max_rate_step_bf_roll()
 {
-    float alpha = MIN(get_rate_roll_pid().get_filt_E_alpha(), get_rate_roll_pid().get_filt_D_alpha());
+    float alpha = MIN(get_rate_roll_pid().get_filt_E_alpha(AP::scheduler().get_filtered_loop_time()), get_rate_roll_pid().get_filt_D_alpha(AP::scheduler().get_filtered_loop_time()));
```
Comment (IamPete1) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:1070]: Here we change to using the filtered loop time?

**Review by IamPete1 — COMMENTED** [review]
We have three "times" used in different places. I guess the time most should be running on is the time between the last ins cycles, which I guess we assume to be the loop time. Some places use loop time and others use filtered loop time. This also uses the ins read time directly in a few places. 

I'm never really sure what time we should use for PIDs, should it be the time between sensor samples or motor outputs. On copter where there done as the first thing each loop I guess there is very little difference. However on plane we run the attitude control at the end of the loop thus introducing a fair bit of variation. 

Basically I'm not 100% sure what time we actually want, and what time we should use in AP as a best approximation. Once we have that approximation I would have expected to use that everywhere but here is a bit of a mixture. I guess the best might depend on the level of the control loop.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp (hunk 2 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_PosControl.cpp:519]: This is where we are checking that we have initialised the position controller on the last loop or we were running on the last loop. This is why this is different.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp (hunk 1 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:241]: This is where I get stupid but I figured that we were just accessing an existing and pre-calculated value. This form also represents what we are doing more accuratly I think. This is just my first go for discussion though so I am open to all and any suggestions and discussions.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp (hunk 2 above)
Comment (andyp1per) [reply libraries/AC_AttitudeControl/AC_PosControl.cpp:519]: Do you think we want this or _last_sample_us?

**Comment by lthall** [comment]
Yes, we use different approaches on the different aircraft types. I have built the copter based on what I think we should do. The others I have built based on the minimum change as they tend to be less processor heavy and I have less, less understanding of the code and less insight  into the lead developers thoughts.

I am sure we will need some discussion and refinement before we settle on the final implementation that is hopefully pretty uniform accross all aircraft types.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp (hunk 1 above)
Comment (andyp1per) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:241]: I suspect it will all get inlined anyway

File: libraries/AC_AttitudeControl/AC_PosControl.cpp (hunk 2 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_PosControl.cpp:519]: This is effectivly the replacement of the flow of control error. We just want to make sure we have initialised the Pos controller. So strictly speeking we just need to make sure that AP::ins().get_last_update_usec() - _last_update_z_us = get_dt()* 1000000.0 but I doubled it to give it a buffer.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp (hunk 3 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:1070]: Yes, this function attempts to predict what the next loop will be so we want the average loop time or the target loop time. I am not sure if I chose the correct one yet, I need to think about this a little more.

**Comment by tridge** [comment]
should be get_last_loop_time_s() to get the intra-gyro time

**Comment by peterbarker** [comment]
I wonder if this might end up fixing https://github.com/ArduPilot/ardupilot/issues/10562

**Comment by rmackay9** [comment]
We discussed this on the EU dev call and decided that we should try to backport this to 4.3 in order to improve performance for those users with F4 CPUs.

Tridge mentioned that he would like to see testing to confirm that the existing tune works at different loop rates and variable loop rights (e.g. the main loop rates slows or speeds up in flight).

File: libraries/AC_PID/AC_P_1D.cpp [hunk 4]
```diff
@@ -38,7 +37,7 @@ float AC_P_1D::update_all(float &target, float measurement)
     }
 
     // MIN(_Dxy_max, _D2xy_max / _kxy_P) limits the max accel to the point where max jerk is exceeded
-    return sqrt_controller(_error, _kp, _D1_max, _dt);
+    return sqrt_controller(_error, _kp, _D1_max, 0.0);
```
Comment (rmackay9) [inline libraries/AC_PID/AC_P_1D.cpp:40]: What effect does hardcoding the dt argument here have?  I see this is done in both AC_P_1D and AC_P_2D.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp [hunk 5]
```diff
@@ -583,10 +587,10 @@ void AC_PosControl::stop_vel_xy_stabilisation()
     _pid_vel_xy.reset_I();
 }
 
-// is_active_xy - returns true if the xy position controller has been run in the previous 5 loop times
+// is_active_xy - returns true if the xy position controller has bee n run in the previous 5 loop times
 bool AC_PosControl::is_active_xy() const
 {
-    return ((AP_HAL::micros64() - _last_update_xy_us) <= _dt * 5000000.0);
+    return ((AP::ins().get_last_update_usec() - _last_update_xy_us) <= _dt* 2000000.0);
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp:593]: Why make this more strict?

File: libraries/AC_AttitudeControl/AC_PosControl.h [hunk 6]
```diff
@@ -12,6 +12,7 @@
 #include <AC_PID/AC_PID_2D.h>       // PID library (2-axis)
 #include <AP_InertialNav/AP_InertialNav.h>  // Inertial Navigation library
 #include "AC_AttitudeControl.h"     // Attitude control library
+#include <AP_InertialSensor/AP_InertialSensor.h>
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_PosControl.h:15]: No big deal but I wonder if we really need to include AP_InertialSensor.h.  Certainly we need it in the .cpp but I don't think we need it here in the .h.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp [hunk 7]
```diff
@@ -901,7 +904,7 @@ void AC_PosControl::update_pos_offset_z(float pos_offset_z)
 // is_active_z - returns true if the z position controller has been run in the previous 5 loop times
 bool AC_PosControl::is_active_z() const
 {
-    return ((AP_HAL::micros64() - _last_update_z_us) <= _dt * 5000000.0);
+    return ((AP::ins().get_last_update_usec() - _last_update_z_us) <= _dt* 2000000.0);
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp:907]: replacing the AP_HAL::micros with the last INS update time is a nice enhancement but is actually unrelated to the main purpose of this PR.   We have millions of places where we make unnecessary calls to AP_HAL to ask the time.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp [hunk 8]
```diff
@@ -1067,7 +1068,8 @@ float AC_AttitudeControl::get_althold_lean_angle_max_cd() const
 // Return roll rate step size in centidegrees/s that results in maximum output after 4 time steps
 float AC_AttitudeControl::max_rate_step_bf_roll()
 {
-    float alpha = MIN(get_rate_roll_pid().get_filt_E_alpha(), get_rate_roll_pid().get_filt_D_alpha());
+    float dt_average = AP::scheduler().get_filtered_loop_time();
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:1071]: As a side note it is slightly annoying that AP_Scheduler::get_filtered_loop_time() doesn't have a comment above it (all other methods in the class do have comments) so it takes extra digging to ensure it is really providing the results in seconds.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.h [hunk 9]
```diff
@@ -74,6 +72,12 @@ class AC_AttitudeControl {
     // Empty destructor to suppress compiler warning
     virtual ~AC_AttitudeControl() {}
 
+    // set_dt / get_dt - dt is the time since the last time the attitude controllers were updated
+    // _dt should be set based on the time of the last IMU read used by these controllers
+    // the attitude controller should run updates for active controllers on each loop to ensure normal operation
+    void set_dt(float dt) { _dt = dt; }
+    float get_dt() const { return _dt; }
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.h:79]: I don't immediately see any places where we call AC_AttitudeControl::get_dt().  I do see calls to AC_PosControl::get_dt().

It's again not a blocker but we don't normally add methods that aren't used.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp (hunk 8 above)
Comment (rmackay9) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:1071]: This use of the filtered_loop_time() seems to be the most significant change in AC_PosControl.

File: ArduCopter/Attitude.cpp [hunk 10]
```diff
@@ -1,5 +1,21 @@
...
+/*************************************************************
+ *  Attitude Rate controllers and timing
+ ****************************************************************/
+
+// update rate controllers and output to roll, pitch and yaw actuators
+//  called at 400hz by default
+void Copter::run_rate_controller()
+{
+    // set attitude and position controller loop time
+    attitude_control->set_dt(AP::scheduler().get_last_loop_time_s());
```
Comment (rmackay9) [inline ArduCopter/Attitude.cpp:12]: It might be slightly better style to have a, "const float last_loop_time_s = AP::scheduler().get_last_loop_time_s()" and then pass this into the next two functions instead of calling the scheduler twice.  It should make no functional difference of course.

Also there's perhaps no need to use "AP::scheduler()." instead simply "scheduler." could be used.

File: libraries/AP_Math/AP_Math.cpp [hunk 11]
```diff
@@ -399,9 +399,13 @@ Vector3F get_vel_correction_for_sensor_offset(const Vector3F &sensor_offset_bf,
  */
 float calc_lowpass_alpha_dt(float dt, float cutoff_freq)
 {
-    if (dt <= 0.0f || cutoff_freq <= 0.0f) {
+    if (is_negative(dt) || !is_positive(cutoff_freq)) {
+        INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
```
Comment (rmackay9) [inline libraries/AP_Math/AP_Math.cpp:403]: I think we should use, "AP_InternalError::error_t::invalid_arg_or_result".  The result won't be any different but this might be slightly closer to describing what is wrong.

File: libraries/Filter/SlewLimiter.cpp [hunk 12]
```diff
@@ -40,6 +40,10 @@ SlewLimiter::SlewLimiter(const float &_slew_rate_max, const float &_slew_rate_ta
  */
 float SlewLimiter::modifier(float sample, float dt)
 {
+    if(!is_positive(dt)) {
```
Comment (rmackay9) [inline libraries/Filter/SlewLimiter.cpp:43]: maybe put a space between, "if" and "("?

**Comment by rmackay9** [comment]
This looks good to me.

Beyond answering/commenting on issues raised by other devs, I think all that remains is testing.  These vehicles (at least) are affected so the vehicle maintainers should probably comment whether they are happy for these changes to go in.

Multicopters
Heli
QuadPlane
Blimb (especially loiter and velocity mode)
Sub
Antenna Tracker

**Review by tridge — CHANGES_REQUESTED** [review]
on CubeOrange quadplane I'm getting:
AP: PreArm: Internal errors 0x100000 l:403 flow_of_ctrl
that is in float calc_lowpass_alpha_dt(float dt, float cutoff_freq)

**Comment by tridge** [comment]
hit the internal error on copter too, needed to change:
ATC_RAT_RLL_FLTE and ATC_RAT_PIT_FLTE to non-zero

**Comment by tridge** [comment]
@lthall we did some flights at increasing loop rates on a 7" quad, going from 400Hz to 2000Hz in steps of 400Hz
Logs here: http://uav.tridgell.net/tmp/loop_times/
note that we lose quite a few log msgs at high loop rates

File: libraries/AC_AttitudeControl/AC_AttitudeControl.h (hunk 9 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.h:79]: Yes, there was a place I needed to call this instead of the Position controller.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp (hunk 8 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:1071]: This only impacts the autotune step size.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp (hunk 7 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_PosControl.cpp]: Not really. AP::ins().get_last_update_usec() is the equivlent of AP_HAL::micros64() given we are referencing the time we run this controller to the last gyro read. In any normal cycle we expect:
(AP::ins().get_last_update_usec() - _last_update_z_us) = _dt

So any increase over this means we have missed running this in a cycle. I have increased this back up to 5 times but I think it should always be ==

File: libraries/AC_AttitudeControl/AC_PosControl.cpp (hunk 5 above)
Comment (lthall) [reply libraries/AC_AttitudeControl/AC_PosControl.cpp]: Yes backed it off again.

File: libraries/AC_PID/AC_P_1D.cpp (hunk 4 above)
Comment (lthall) [reply libraries/AC_PID/AC_P_1D.cpp:40]: When using the square root controller for shaping we precisly follow this path to zero and get an occilation. However when using this controller in a PID loop the response is only aproximate and also noisy. So this dt is only important when using the square root for ideal shaping or comand model operations, not control operations.

**Comment by MichelleRos** [comment]
We did another flight on the same copter. In acro for most of the flight but when switching to loiter at the end it was again very jittery.
https://drive.google.com/file/d/1xfkK5rVJeDgJ-soOMkQL2-tGcR5sPq3V/view?usp=sharing

**Comment by MichelleRos** [comment]
We did another flight on the same copter. In acro for most of the flight but when switching to loiter at the end it was again very jittery.
https://drive.google.com/file/d/1xfkK5rVJeDgJ-soOMkQL2-tGcR5sPq3V/view?usp=sharing

**Comment by lthall** [comment]
I have finally found what looks like the problem with Loiter:
image

Now I just need to work out what I am seeing.....

**Comment by tridge** [comment]
@MichelleRos we should have set LOG_FILE_RATEMAX=400 so we didn't lose so many log msgs

**Comment by lthall** [comment]
Left over questions are:
Best way to handle the is_active check in the position controller (maybe we should put something similar in the attitude controller).
Sub needed to have the motors time set before init to work for some reason.

**Comment by tridge** [comment]
I think the is_active internal errors should be SITL only

File: libraries/AP_Motors/AP_MotorsHeli_Single.cpp [hunk 13]
```diff
@@ -582,31 +582,31 @@ void AP_MotorsHeli_Single::output_to_motors()
...
         (_servo_test_cycle_time >= 6.0f && _servo_test_cycle_time < 6.5f)){
-        _pitch_test += (1.0f / (_loop_rate / 2.0f));
-        _oscillate_angle += 8 * M_PI / _loop_rate;
+        _pitch_test += 2.0 * _dt;
+        _oscillate_angle += 8 * M_PI * _dt;
         _yaw_test = 0.5f * sinf(_oscillate_angle);
     } else if ((_servo_test_cycle_time >= 0.5f && _servo_test_cycle_time < 4.5f)||                            // Roll swash around
                (_servo_test_cycle_time >= 6.5f && _servo_test_cycle_time < 10.5f)){
-        _oscillate_angle += M_PI / (2 * _loop_rate);
+        _oscillate_angle += _dt * M_PI / 2;
```
Comment (rmackay9) [inline libraries/AP_Motors/AP_MotorsHeli_Single.cpp:594]: I guess you might as well make this the same as the equivalent in AP_MotorsHeli_Single. 

_oscillate_angle += M_PI * 0.5 * _dt;

File: libraries/AP_Motors/AP_MotorsMatrix.cpp [hunk 14]
```diff
@@ -414,7 +414,7 @@ void AP_MotorsMatrix::output_armed_stabilizing()
 void AP_MotorsMatrix::check_for_failed_motor(float throttle_thrust_best_plus_adj)
 {
     // record filtered and scaled thrust output for motor loss monitoring purposes
-    float alpha = 1.0f / (1.0f + _loop_rate * 0.5f);
+    float alpha = _dt / (_dt + 0.5f);
```
Comment (rmackay9) [inline libraries/AP_Motors/AP_MotorsMatrix.cpp:417]: This change results in about double the alpha value when loop_rate = 400.  Is this an intentional change?

Comment (lthall) [reply libraries/AP_Motors/AP_MotorsMatrix.cpp:417]: lr =  400
>> 1/(1+lr*0.5)
ans =  0.0050
>> dt = 1/lr
dt =  0.0025
>> dt / (dt + 0.5)
ans =  0.0050

I think this one is right.

Comment (rmackay9) [reply libraries/AP_Motors/AP_MotorsMatrix.cpp:417]: I'm sorry, you're right.  I had a typo when I did my own check.

**Review by tridge — APPROVED** [review]

**Review by rmackay9 — APPROVED** [review]
I've gone through this line by line and I think it's OK.

**Comment by rmackay9** [comment]
This is included in 4.3.3-rc1 (backport was taken from this branch https://github.com/rmackay9/rmackay9-ardupilot/commits/copter-432-rc1-with-dt)
